diff -uNr cgit-0.10.old/cgit.mk cgit-0.10/cgit.mk
--- cgit-0.10.old/cgit.mk	2014-01-17 16:09:16.000000000 +0100
+++ cgit-0.10/cgit.mk	2014-02-25 18:26:04.391520820 +0100
@@ -29,30 +29,18 @@
 	LUA_MESSAGE := linking without specified Lua support
 	CGIT_CFLAGS += -DNO_LUA
 else
-LUAJIT_CFLAGS := $(shell pkg-config --cflags luajit 2>/dev/null)
-LUAJIT_LIBS := $(shell pkg-config --libs luajit 2>/dev/null)
-LUA_LIBS := $(shell pkg-config --libs lua 2>/dev/null)
-LUA_CFLAGS := $(shell pkg-config --cflags lua 2>/dev/null)
-ifeq (JIT,$(LUA_IMPLEMENTATION))
-	ifeq ($(strip $(LUAJIT_LIBS)),)
- 		$(error LuaJIT specified via LUA_IMPLEMENTATION=JIT, but library could not be found.)
-	endif
-	LUA_MESSAGE := linking with selected LuaJIT
-	CGIT_LIBS += $(LUAJIT_LIBS)
-	CGIT_CFLAGS += $(LUAJIT_CFLAGS)
-else ifeq (VANILLA,$(LUA_IMPLEMENTATION))
-	ifeq ($(strip $(LUA_LIBS)),)
- 		$(error Lua specified via LUA_IMPLEMENTATION=VANILLA, but library could not be found.)
-	endif
-	LUA_MESSAGE := linking with selected Lua
-	CGIT_LIBS += $(LUA_LIBS)
-	CGIT_LIBS += $(LUA_CFLAGS)
-else ifneq ($(strip $(LUAJIT_LIBS)),)
-	LUA_MESSAGE := linking with autodetected LuaJIT
-	CGIT_LIBS += $(LUAJIT_LIBS)
-	CGIT_CFLAGS += $(LUAJIT_CFLAGS)
-else ifneq ($(strip $(LUA_LIBS)),)
-	LUA_MESSAGE := linking with autodetected Lua
+ifeq ($(LUA_PKGCONFIG),)
+	LUA_PKGCONFIG := $(shell for pc in lua5.2 lua51 lua50; do \
+			pkg-config --exists $$pc 2>/dev/null && echo $$pc && break; \
+			done)
+	LUA_MODE := autodetected
+else
+	LUA_MODE := specified
+endif
+ifneq ($(LUA_PKGCONFIG),)
+	LUA_MESSAGE := linking with $(LUA_MODE) $(LUA_PKGCONFIG)
+	LUA_LIBS := $(shell pkg-config --libs $(LUA_PKGCONFIG) 2>/dev/null)
+	LUA_CFLAGS := $(shell pkg-config --cflags $(LUA_PKGCONFIG) 2>/dev/null)
 	CGIT_LIBS += $(LUA_LIBS)
 	CGIT_CFLAGS += $(LUA_CFLAGS)
 else
