diff -uNr cgit-0.10.1.old/cgit.mk cgit-0.10.1/cgit.mk
--- cgit-0.10.1.old/cgit.mk	2014-02-28 00:12:08.000000000 +0100
+++ cgit-0.10.1/cgit.mk	2014-03-01 19:02:59.424624275 +0100
@@ -30,7 +30,7 @@
 	CGIT_CFLAGS += -DNO_LUA
 else
 ifeq ($(LUA_PKGCONFIG),)
-	LUA_PKGCONFIG := $(shell for pc in luajit lua lua5.2 lua5.1; do \
+	LUA_PKGCONFIG := $(shell for pc in lua5.2 lua51 lua50; do \
 			pkg-config --exists $$pc 2>/dev/null && echo $$pc && break; \
 			done)
 	LUA_MODE := autodetected
