diff --git a/cgit.mk b/cgit.mk
index 8af0041..35beb03 100644
--- a/cgit.mk
+++ b/cgit.mk
@@ -87,5 +87,6 @@ $(CGIT_PREFIX)CGIT-CFLAGS: FORCE
 $(CGIT_OBJS): %.o: %.c GIT-CFLAGS $(CGIT_PREFIX)CGIT-CFLAGS $(missing_dep_dirs)
 	$(QUIET_CC)$(CC) -o $*.o -c $(dep_args) $(ALL_CFLAGS) $(EXTRA_CPPFLAGS) $(CGIT_CFLAGS) $<
 
-$(CGIT_PREFIX)cgit: $(CGIT_OBJS) GIT-LDFLAGS $(GITLIBS)
+LIBS = -lgit -lgit_xdiff $(EXTLIBS)
+$(CGIT_PREFIX)cgit: $(CGIT_OBJS) GIT-LDFLAGS
 	$(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) $(LIBS)
