--- cgit-0.9.1/ui-log.c~	2012-11-15 01:28:59.000000000 +0100
+++ cgit-0.9.1/ui-log.c	2013-05-17 13:15:53.682385508 +0200
@@ -195,9 +195,9 @@
 				strbuf_addstr(&msgbuf, info->msg);
 				strbuf_addch(&msgbuf, '\n');
 			}
-			format_note(NULL, commit->object.sha1, &msgbuf,
+			format_display_notes(commit->object.sha1, &msgbuf,
 			            PAGE_ENCODING,
-			            NOTES_SHOW_HEADER | NOTES_INDENT);
+			            0);
 			strbuf_addch(&msgbuf, '\n');
 			strbuf_ltrim(&msgbuf);
 		}
--- cgit-0.9.1/ui-commit.c~	2013-05-17 13:09:43.468524461 +0200
+++ cgit-0.9.1/ui-commit.c	2013-05-17 13:17:12.059843403 +0200
@@ -36,7 +36,7 @@
 	}
 	info = cgit_parse_commit(commit);
 
-	format_note(NULL, sha1, &notes, PAGE_ENCODING, 0);
+	format_display_notes(sha1, &notes, PAGE_ENCODING, 1);
 
 	load_ref_decorations(DECORATE_FULL_REFS);
 
